src/api.rs

```diff
@@ -0,0 +1,243 @@
+}
+
+pub async fn invalidate_tokens(user: &User) {
+    Auth::invalidate_tokens(AUTH.clone(), user.id).await
+}
+
+pub async fn get_user_stripped(id: ID) -> Result<GenericUser> {
+    if let Ok(other) = User::load(&id).await {
+        Ok(other.to_generic())
+    } else {
+        Err(Error::UserNotFound)
+    }
+}
+
+pub async fn change_username(user: &mut User, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    let old_name = user.username.clone();
+    user.username = new_name;
+    user.save().await?;
+    Ok(old_name)
+}
+
+pub async fn create_hub(owner: &mut User, name: String) -> Result<ID> {
+    is_valid_username(&name)?;
+    let mut id = new_id();
+    while Hub::load(&id).await.is_ok() {
+        id = new_id();
+    }
+    let mut new_hub = Hub::new(name, id.clone(), &owner);
+    let channel_id = new_hub.new_channel(&owner.id, "chat".to_string()).await?;
+    if let Some(group) = new_hub.groups.get_mut(&new_hub.default_group) {
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::ViewChannel,
+            crate::permission::PermissionSetting::TRUE,
+        );
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::SendMessage,
+            crate::permission::PermissionSetting::TRUE,
+        );
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::ReadMessage,
+            crate::permission::PermissionSetting::TRUE,
+        );
+    }
+    owner.in_hubs.push(id.clone());
+    owner.save().await?;
+    new_hub.save().await?;
+    Ok(id)
+}
+
+pub async fn get_hub(user: &User, hub_id: &ID) -> Result<Hub> {
+    user.in_hub(&hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    hub.strip(&user.id)
+}
+
+pub async fn delete_hub(user: &User, hub_id: &ID) -> Result<()> {
+    user.in_hub(&hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_all_permissions() {
+        if let Ok(()) = tokio::fs::remove_file(hub.get_info_path()).await {
+            if let Ok(()) = tokio::fs::remove_dir_all(hub.get_data_path()).await {
+                Ok(())
+            } else {
+                Err(Error::DeleteFailed)
+            }
+        } else {
+            Err(Error::DeleteFailed)
+        }
+    } else {
+        Err(Error::NoPermission)
+    }
+}
+
+pub async fn rename_hub(user: &User, hub_id: &ID, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    user.in_hub(&hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_permission(HubPermission::Administrate, &hub) {
+        let old_name = hub.name.clone();
+        hub.name = new_name;
+        hub.save().await?;
+        Ok(old_name)
+    } else {
+        Err(Error::NoPermission)
+    }
```

Comment:
You might want to simplify permission checks even more since you are essentially writing return Error::NoPermission over and over again. A function with a closure might be good enough, or you can go with a macro.

Reply:
That's a good point, I was originally going to go with a `Error::NoPermission(PermissionType)` but then I realized that I had two permission types, (Hub and Channel), if I decide not to do that then I will probably replace these types of statements with a macro of some sort...

Reply:
I ended up combining those too (macro but also specify which permission is missing)